subprojects/language-java/src/main/java/org/gradle/api/tasks/JavaExec.java

```diff
@@ -131,18 +130,37 @@ public JavaExec() {
         javaExecSpec.getMainClass().convention(mainClass);
         javaExecSpec.getMainModule().convention(mainModule);
         javaExecSpec.getModularity().getInferModulePath().convention(modularity.getInferModulePath());
-        javaLauncher = objectFactory.property(JavaLauncher.class);
+
+        JavaToolchainService javaToolchainService = getJavaToolchainService();
+        Provider<JavaLauncher> javaLauncherConvention = getProviderFactory()
+            .provider(() -> JavaExecExecutableUtils.getExecutableOverrideToolchainSpec(this, objectFactory))
```

Comment:
I wonder if it might be better to pass a `ProviderFactory` to the `JavaExecExecutableUtils` and move all the work of creating this `Provider` into a `Provider<JavaLauncher> getJavaLauncherConvention()` method that lives there?

Reply:
That is a good suggestion. We'll try to polish these places in the follow-ups 👍